checks.d/kafka_consumer.py

```diff
@@ -29,38 +29,92 @@ def __init__(self, name, init_config, agentConfig, instances=None):
         self.kafka_timeout = int(
             init_config.get('kafka_timeout', DEFAULT_KAFKA_TIMEOUT))
 
+    def _get_children(self, zk_conn, zk_path, name_for_error):
+        """Fetch child nodes for a given Zookeeper path"""
+        children = []
+        try:
+            children = zk_conn.get_children(zk_path)
+        except NoNodeError:
+            self.log.warn('No zookeeper node at %s' % zk_path)
+        except:
+            self.log.exception('Could not read %s from %s' % (name_for_error, zk_path))
+        return children
+
     def check(self, instance):
-        consumer_groups = self.read_config(instance, 'consumer_groups',
-                                           cast=self._validate_consumer_groups)
+        # Only validate consumer_groups if specified; fallback to zk otherwise
+        consumer_groups_specified = True if 'consumer_groups' in instance else False
+        if consumer_groups_specified:
+            consumer_groups = self.read_config(instance, 'consumer_groups',
+                                               cast=self._validate_consumer_groups)
+
         zk_connect_str = self.read_config(instance, 'zk_connect_str')
         kafka_host_ports = self.read_config(instance, 'kafka_connect_str')
 
         # Construct the Zookeeper path pattern
+        # /consumers/[groupId]/offsets/[topic]/[partitionId]
         zk_prefix = instance.get('zk_prefix', '')
-        zk_path_tmpl = zk_prefix + '/consumers/%s/offsets/%s/%s'
+        zk_path_consumer = zk_prefix + '/consumers/'  # /consumers/
+        zk_path_topic_tmpl = zk_path_consumer + '%s/offsets/'  # /consumers/[groupID]/offsets/
+        zk_path_partition_tmpl = zk_path_topic_tmpl + '%s/'  # /consumers/[groupID]/offsets/[topic]/
 
         # Connect to Zookeeper
         zk_conn = KazooClient(zk_connect_str, timeout=self.zk_timeout)
         zk_conn.start()
 
+        # We need to track the topics/partitions for which we're measuring
+        # consumer offset so that we can lookup the partition's newest message
+        # in the Kafka cluster to calculate consumer lag.
+        # This is tricky because different consumer groups may track different
+        # partitions of a topic. So when we modify a topic's list of partitions,
+        # we don't want to overwrite, only add missing partitions. And only want
+        # uniques, because even though each partition may have multiple consumer
+        # group offsets, it will only have one broker offset. Hence the
+        # defaultdict(set) usage.
+        # {topic1: set[(partition1, partition2)], topic2: set[partition1]}
+        tracked_topic_partitions = defaultdict(set)
```

Comment:
So I actually did try this originally, as I had the exact same thought (`defaultdict` came from the current code).

However, I ran into problems without it. My memory is a little hazy as it's been a few weeks, but IIRC it's needed to pre-populate the set in the dict so that L#106 updates rather than overwrites:

    tracked_topic_partitions[topic].update(set(partitions))